### PR TITLE
Treant.js formatted JSON from parse tree representation

### PR DESCRIPTION
**Pull request by bisakhmondal**
input:
```json
{
    "S": {
        "A": {
            "a": null
        },
        "B": {
            "b": null
        },
        "A_1": {
            "a": null
        }
    }
}
```

Output:
```json
{
    "text": {
        "name": "S"
    },
    "children": [
        {
            "text": {
                "name": "A"
            },
            "children": [
                {
                    "text": {
                        "name": "a"
                    }
                }
            ]
        },
        {
            "text": {
                "name": "B"
            },
            "children": [
                {
                    "text": {
                        "name": "b"
                    }
                }
            ]
        },
        {
            "text": {
                "name": "A_1"
            },
            "children": [
                {
                    "text": {
                        "name": "a"
                    }
                }
            ]
        }
    ]
}
```

### CODE REVIEW

File: src/parser/parser.cpp [hunk 1]
```diff
@@ -166,8 +166,56 @@ bool Parser::WriteParseTree(const std::string &filepath) {
...
-
   return false;
 }
 
+/**
+ * Supportive function for Parser::FormattedJSON
+ * @param value
+ * @return { text: { name: "value" } }
+ */
+json GetTextNode(const std::string &value) {
```
Comment (noob77777) [inline src/parser/parser.cpp]: Make utility functions private members?
Just a suggestion

File: test/parser/parser_test.cpp [hunk 2]
```diff
@@ -363,4 +363,35 @@ TEST(parser, Parser4) {
...
+   *          }
+   *     ]
+   * }
+   */
+  json formatted = Parser::FormattedJSON(tree);
+  ASSERT_EQ(formatted.dump(),
+            "{\"text\":{\"name\":\"S\"},\"children\":[{\"text\":{\"name\":\"A\"},\"children\":[{\"text\":{\"name\":"
+            "\"a\"}}]},{\"text\":{\"name\":\"B\"},\"children\":[{\"text\":{\"name\":\"b\"}}]},{\"text\":{\"name\":\"A_"
+            "1\"},\"children\":[{\"text\":{\"name\":\"a\"}}]}]}");
+  std::cout << tree.dump(4) << "\n";
```
Comment (noob77777) [inline test/parser/parser_test.cpp]: Remove couts

File: src/parser/parser.cpp [hunk 3]
```diff
@@ -166,8 +166,56 @@ bool Parser::WriteParseTree(const std::string &filepath) {
...
+  }
+  return "";
+}
+
+/**
+ * Utility recursive function for Parser::FormattedJSON
+ * @param body, a json
+ * @returns Treant.js formatted JSON
+ */
+json RecRunner(json main, std::string key = "") {
```
Comment (noob77777) [inline src/parser/parser.cpp]: Make const &

File: src/parser/parser.cpp (hunk 1 above)
Comment (bisakhmondal) [reply src/parser/parser.cpp]: I think for a single case it does need to be a member at all!!

File: test/parser/parser_test.cpp (hunk 2 above)
Comment (bisakhmondal) [reply test/parser/parser_test.cpp]: ya. thanks

File: src/parser/parser.cpp (hunk 3 above)
Comment (bisakhmondal) [reply src/parser/parser.cpp]: Wouldn't make much of a difference from a memory perspective.
Inside the body, we have to create a copy of this constant to handle base cases.

File: src/parser/parser.cpp (hunk 1 above)
Comment (noob77777) [reply src/parser/parser.cpp]: make 3 utility functions private static members i guess. looks better

Comment (bisakhmondal) [reply src/parser/parser.cpp]: okay, sounds good

File: codecov.yml [hunk 4]
```diff
@@ -12,7 +12,7 @@ coverage:
         threshold: 5%
     patch:
       default:
-        target: 85%
```
Comment (noob77777) [inline codecov.yml]: 85%

File: src/parser/parser.cpp [hunk 5]
```diff
@@ -166,8 +166,57 @@ bool Parser::WriteParseTree(const std::string &filepath) {
...
+  j["text"]["name"] = value;
+  return j;
+}
+
+/**
+ * Supportive function for Parser::FormattedJSON
+ * @param value
+ * @return the current key of the JSON head
+ */
+std::string GetHeadKey(const json &j) {
```
Comment (noob77777) [inline src/parser/parser.cpp]: this is not used. remove if not required.

**Review by noob77777 — CHANGES_REQUESTED** [review]

File: src/parser/parser.cpp [hunk 6]
```diff
@@ -166,8 +166,57 @@ bool Parser::WriteParseTree(const std::string &filepath) {
...
+  }
+
+  json j = GetTextNode(key);
+  for (auto it = body.begin(); it != body.end(); it++) {
+    j["children"].push_back(RecRunner(it.value(), it.key()));
+  }
+  return j;
+}
+
+json Parser::FormattedJSON(const json &body) { return RecRunner(body); }
```
Comment (noob77777) [inline src/parser/parser.cpp]: use  GetHeadKey and remove default param in RecRunner? just a suggestion.

File: src/parser/parser.cpp [hunk 7]
```diff
@@ -166,8 +166,57 @@ bool Parser::WriteParseTree(const std::string &filepath) {
     ofs << parse_tree_.dump(INDENTATION) << '\n';
```
Comment (noob77777) [inline src/parser/parser.cpp]: update this to include formmatted output.

File: src/parser/parser.cpp (hunk 5 above)
Comment (bisakhmondal) [reply src/parser/parser.cpp]: yes, resolved

File: src/parser/parser.cpp (hunk 1 above)
Comment (bisakhmondal) [reply src/parser/parser.cpp]: done

**Review by noob77777 — DISMISSED** [review]

**Review by noob77777 — APPROVED** [review]

**Pull request merged**